Add Blockchain Integration Framework documentation (#1038)
smartcontractkit/documentation: docs/blockchain-integrations-framework/index.md

**Test compatibility of the network client API:**

```shell
bif compatibility test-rpc \ 
--privateKey <PRIVATE_KEY> \
--chainID <CHAIN_ID> \
--rpc <RPC_URL> \
--storageContractAddress <OPTIONAL_CONTRACT_ADDRESS>
```

The `--storageContractAddress` flag is available so you can re-use deployed contracts from previous tests and save wallet funds.

**Test the smart contract op codes of a network:**

```
bif compatibility test-opcodes \
--privateKey <PRIVATE_KEY> \
--chainID <CHAIN_ID> \
--rpc <RPC_URL> \
--opcodesContractAddress <OPTIONAL_CONTRACT_ADDRESS>
```

The `--opcodesContractAddress` flag is available so you can re-use deployed contracts from previous tests and save wallet funds.

## Running integration tests

Integration testing uses the `bif integration test` command and a TOML file. The TOML file simplifies managing the many variables involved.

The TOML file for the Blockchain Integrations Framework has two settings sections for OCR soak tests on EVM networks:

- Network settings: Variables that configure which chain you use for testing and the private key for the wallet you want to use for deploying contracts

> - Test settings: Test duration, node funding, and round configuration settings to control the behavior of the test

Reviewer: ```suggestion
- *Test settings*: Test duration, node funding, and round configuration settings to control the behavior of the test
```

---

Chainlink Labs provides self-serve tools that node operators and blockchain projects can use to deploy Data Feed integrations on EVM networks. These tools are called the Blockchain Integrations Framework (BIF). This framework includes a command line interface (CLI) and is available as a binary file. The binary is not publicly available. During the Alpha release, you must receive this binary directly from Chainlink Labs.

The framework simplifies the following tasks:

- Deploy Chainlink nodes into Kubernetes clusters and distribute jobs to them
- Run tests that generate reports detailing compatibility between Chainlink Labs' EVM integration and a given blockchain
- Deploy cluster infrastructure required to test OCR for Data Feeds
- Perform longer running soak tests for OCR for Data Feeds
- Perform smart contract management tasks with the same capabilities as Chainlink Labs' tools

**Contents**

- [Requirements](#requirements)
- [Set up your environment](#set-up-your-environment)
- [Running basic tests](#running-basic-tests)
- [Running integration testing](#running-integration-tests)
- [Troubleshooting integration tests](#troubleshooting-integration-tests)

## Requirements

> Most basic tests can be completed on a system with minimal hardware resources. To complete soak testing and use the CLI to create clusters with multiple Chainlink nodes, you must run a Kubernets cluster on a system with the following system resources:

Reviewer 2: ```suggestion
Most basic tests can be completed on a system with minimal hardware resources. To complete soak testing and use the CLI to create clusters with multiple Chainlink nodes, you must run a Kubernetes cluster on a system with the following system resources:
```

> Most basic tests can be completed on a system with minimal hardware resources. To complete soak testing and use the CLI to create clusters with multiple Chainlink nodes, you must run a Kubernets cluster on a system with the following system resources:

Reviewer 2: I think we should mention the Docker requirement here, and say that Docker must be allotted at least 6 CPU and 10 GB RAM.

Author: I added Docker here and also in the environment setup instructions with a link to https://docs.docker.com/engine/install/

How do you configure Docker default limits? I'm not seeing it in the docs, and it looks like Docker uses all available CPUs and memory by default: https://docs.docker.com/config/containers/resource_constraints/#cpu

Reviewer 2: By default, any Docker container is given unlimited access to the resources available to Docker. But Docker may only be allotted 4CPUs and 8GB RAM, which would be insufficient for our purposes. I haven't found a way to boost these limits without Docker Desktop. Here are the Linux instructions: https://docs.docker.com/desktop/settings/linux/

Author: Added in the most recent commit. I added info in both the Resources section and the environment setup instructions.

- A Linux system
- 6 CPU cores
- 10 GiB of system memory
- At least 50 GiB of disk storage

## Set up your environment

Install the framework and create an alias:

1. Obtain the binary file from Chainlink Labs and put it in your local directory.
1. Make the binary available in `/usr/bin` so you can run it without the full path: `sudo mv ./bif /usr/bin/`

You can now run the framework CLI. Run `bif integration -h` to see the available commands. See the [Running basic tests](#running-basic-tests) section for examples.

> If you plan to run soak tests or other tests that require running Chainlink nodes, set up [Helm](https://helm.sh/docs/intro/install/#through-package-managers), [Kubernets](https://kubernetes.io/docs/setup/) and other required tools:

Reviewer 2: ```suggestion
If you plan to run soak tests or other tests that require running Chainlink nodes, set up [Helm](https://helm.sh/docs/intro/install/#through-package-managers), [Kubernetes](https://kubernetes.io/docs/setup/) and other required tools:
```

- The chain ID for your network, which usually can be found on the [LINK Token Contracts](https://docs.chain.link/docs/link-token-contracts/) or [chainlist.org](https://chainlist.org/)
- The RPC URL for either a network node that you run or a third-party service like [Alchemy](https://www.alchemy.com/), [Infura](https://infura.io/), and [Ankr](https://www.ankr.com/rpc/)

**Test compatibility of the network client API:**

```shell
bif compatibility test-rpc \ 
--privateKey <PRIVATE_KEY> \
--chainID <CHAIN_ID> \
--rpc <RPC_URL> \
--storageContractAddress <OPTIONAL_CONTRACT_ADDRESS>
```

The `--storageContractAddress` flag is available so you can re-use deployed contracts from previous tests and save wallet funds.

**Test the smart contract op codes of a network:**

```
bif compatibility test-opcodes \
--privateKey <PRIVATE_KEY> \
--chainID <CHAIN_ID> \
--rpc <RPC_URL> \
--opcodesContractAddress <OPTIONAL_CONTRACT_ADDRESS>
```

The `--opcodesContractAddress` flag is available so you can re-use deployed contracts from previous tests and save wallet funds.

## Running integration tests

> Integration testing uses the `bif integration soak` command and a TOML file. The TOML file simplifies managing the many variables involved.

Reviewer 2: ```suggestion
Integration testing uses the `bif integration test` command and a TOML file. The TOML file simplifies managing the many variables involved.
```

> The TOML file for the Blockchain Integrations Framework has two settings sections:

Reviewer 2: ```suggestion
The TOML file for the Blockchain Integrations Framework has two settings sections for OCR soak tests on EVM networks:
```

- Network settings: Variables that configure which chain you use for testing and the private key for the wallet you want to use for deploying contracts

> - Test settings: Node count and test duration settings to control the behavior of the test

Reviewer 2: Node count is no longer set here, it is handled by infra
```suggestion
- Test settings: Test duration, node funding, and round configuration settings to control the behavior of the test
```

        # Name of the person running the tests (no spaces)
        chainlink_env_user="My-Name" 
        test_log_level="info" # info | debug | trace
        node_count=6
        test_duration=15 # minutes
        contract_count=2
        node_funding=0.1 # ETH
        round_timeout=15 # minutes
        expected_round_time=2 # minutes
        time_between_rounds=1 # minutes

        [tests.v1-7-0.overrides.shorter-test]
        test_duration=2 # minutes

    [tests.v1-9-0]

        [tests.v1-9-0.base]
        keep_environments="Never" # Always | OnFail | Never
        # Image repo to pull the Chainlink image from
        chainlink_image="public.ecr.aws/chainlink/chainlink"
        # Version of the Chainlink image to pull
        chainlink_version="1.9.0"
        # Name of the person running the tests (no spaces)
        chainlink_env_user="My-Name"
        test_log_level="info" # info | debug | trace
        node_count=6  
        test_duration=15 # minutes
        contract_count=2
        node_funding=0.1 # ETH
        round_timeout=15 # minutes
        expected_round_time=2 # minutes
        time_between_rounds=1 # minutes

        [tests.v1-9-0.overrides.shorter-test]

>         test_duration=2 # minutes

Reviewer 2: More up to date TOML, this one contains a `nodes` section that is used by the infra module:
```toml
# example TOML config

# node settings
[nodes]

    [nodes.goerli]

        [nodes.goerli.base]
        eth_url = "wss://goerli"
        eth_http_url = "https://goerli"
        eth_chain_id = "5"

# network settings
[networks]

    [networks.evm]

        [networks.evm.base]
        name = "EVM"
        evm_keys = ["YOUR_EVM_KEY_HERE"]
        ws_urls = []
        chain_id = 1337
        simulated = false
        chainlink_transaction_limit = 5000
        transaction_timeout = "2m"
        minimum_confirmations = 1
        gas_estimation_buffer = 1000
        client_implementation = "Ethereum"

        [networks.evm.overrides.goerli]
        name = "Goerli"
        ws_urls = ["wss://goerli"]
        chain_id = 5
        transaction_timeout = "5m"
        minimum_confirmations = 5

        [networks.evm.overrides.palm]
        name = "Palm"
        ws_urls = ["wss://palm"]
        chain_id = 11297108099

        [networks.evm.overrides.rinkeby]
        name = "Rinkeby"
        ws_urls = ["wss://rinkeby"]
        chain_id = 4

        [networks.evm.overrides.optimism]
        name = "Optimism"
        ws_urls = ["wss://optimism"]
        chain_id = 420
        client_implementation = "Optimism"

# test settings
[tests]

    [tests.v1-7-0]

        [tests.v1-7-0.base]
        keep_environments="Never" # Always | OnFail | Never
        chainlink_image="public.ecr.aws/chainlink/chainlink" # Image repo to pull the Chainlink image from
        chainlink_version="1.7.0-nonroot" # Version of the Chainlink image to pull
        chainlink_env_user="Satoshi-Nakamoto" # Name of the person running the tests (change to your own)
        test_log_level="info" # info | debug | trace
        test_duration=15 # minutes
        contract_count=2
        node_funding=0.1 # ETH
        round_timeout=15 # minutes
        expected_round_time=2 # minutes
        time_between_rounds=1 # minutes

        [tests.v1-7-0.overrides.shorter-test]
        test_duration=2 # minutes
```
You can add a line about the `nodes` section as well, saying that the infra module uses it to configure environment variables on nodes that it deploys.

```

If you completed the full [steps to set up your environment](#set-up-your-environment), you can run a soak test using the following steps:

1. Create a file called `example.toml` and paste the example TOML file above.

> 1. Edit the file to add your `private_keys` under the `networks.evm.base` config.

Reviewer 2: ```suggestion
1. Edit the file to add your `evm_keys` under the `networks.evm.base` config.
```
This was a recent update made in chainlink-testing-framework

1. Modify the `ws_urls` under `networks.evm.overrides.goerli` with your specified RPC endpoint URL. Optionally, you can create additional overrides for different networks`.

> 1. Start running a soak test `bif integration soak ./example.toml`.

Reviewer 2: ```suggestion
1. Start running a soak test `bif integration test ./example.toml`.
```

> 1. The interactive menu opens. Select evm, your desired network overrides, the Chainlink node version, and your test length settings.

Reviewer 2: ```suggestion
1. The interactive menu opens. Select your desired test suite, test, network setting and override, test setting and override, and cluster to run on. For example, `OCR`, `EVMSoak`, `evm`, `goerli`, `v1-7-0` `shorter-test`, and `new cluster`
```

1. While your test is running, view your local Grafana instance at `localhost:3000` to see test metrics.

> 1. Check node logs: `bif integration logs NAMESPACE [--level LOG_LEVEL]`

Reviewer 2: Namespace will be selected via prompt
```suggestion
1. Check node logs: `bif integration logs [--level LOG_LEVEL]`
```

As you test additional networks, modify the TOML file and configure it with the details for your network.

### Troubleshooting integration tests

There are some common issues with soak testing that you can troubleshoot.

**Pods remaining from failed tests:**

If a soak test fails, the pods used for testing often remain running. Use `kubectl` to identify and delete these pods:

1. List all pods by their namespaces: `kubectl get pods --all-namespaces`
1. Delete the deployments that you no longer need by namespace: `kubectl delete --all deployments --namespace=ocr-soak-goerli-50b0c`
1. Modify your test configuration and run `bif integration soak ./example.toml` again.

**Insufficient resources:**

Your system must meet the defined [Requirements](#requirements). Running on systems without sufficient resources will cause soak tests to fail because some containers for Chainlink nodes will not be able to start.

- If you are running on a virtual machine, add additional vCPUs, system memory, or storage.

> - Run your soak tests with fewer nodes. The default value in the TOML config is `node_count=6`, which is recommended for a thorough soak test. You can adjust this value if you are unable to obtain the necessary system resources.

Reviewer 2: Node count is no longer set through TOML. It can be set via `bif infra cluster create --count 6` or `bif integration test --count 6`.
```suggestion
- Run your soak tests with fewer nodes. The default node count is 6 per cluster, which is recommended for a thorough soak test. You can adjust this value if you are unable to obtain the necessary system resources.
```

**Insufficient wallet funding**

> The soak tests send 0.1 ETH to each test node by default, so you must make sure that there are significant funds in the wallet that you use to fund testing. This is the wallet specified by your `private_keys` parameter in the TOML file. Alternatively, you can adjust the `node_funding` parameter in the TOML test configuration, but make sure that each node has enough funds to complete the soak test.

Reviewer 2: ```suggestion
The soak tests send 0.1 ETH to each test node by default, so you must make sure that there are significant funds in the wallet that you use to fund testing. This is the wallet specified by your `evm_keys` parameter in the TOML file. Alternatively, you can adjust the `node_funding` parameter in the TOML test configuration, but make sure that each node has enough funds to complete the soak test.
```
